---
title: "Report: Design of guide RNAs with `snakemake-crispr-guides`"
auhtor: "`r sessionInfo()$platform`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   rmd: "report.Rmd"
output:
  html_document:
    theme: cosmo
    toc: yes
    toc_depth: 2
    number_sections: yes
    df_print: paged
---

----------

# Background

This workflow is a best-practice workflow for the automated generation of guide RNAs for CRISPR applications. It's main purpose is to provide a simple, efficient and easy-to-use framework to design thousands of guides simultaneously for CRISPR libraries from as little input as an organism's name/genome ID. For the manual design of single guides, users are instead referred to even simpler web resources such as [Chop-Chop](http://chopchop.cbu.uib.no/), [CRISPick](https://portals.broadinstitute.org/gppx/crispick/public), or [Cas-OFFinder/Cas-Designer](http://www.rgenome.net/cas-designer/).

This workflow relies to a large degree on the underlying [Bioconductor package ecosystem `crisprVerse`](http://bioconductor.org/packages/release/bioc/html/crisprVerse.html), published in 2022 by:

> Hoberecht, L., Perampalam, P., Lun, A. et al. _A comprehensive Bioconductor ecosystem for the design of CRISPR guide RNAs across nucleases and technologies_. Nat Commun 13, 6568 (**2022**). https://doi.org/10.1038/s41467-022-34320-7.

The workflow is built using [snakemake](https://snakemake.readthedocs.io/en/stable/) and consists of the following steps:

1. Obtain genome database in `fasta` and `gff` format (`python`, [NCBI Datasets](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/))
   1. Using automatic download from NCBI with a `RefSeq` ID
   2. Using user-supplied files
2. Find all possible guide RNAs for the given sequence, with many options for customization (`R`, `crisprVerse`)
3. Collect on-target and off-target scores (`R`, `crisprVerse`, `Bowtie`)
4. Filter and rank guide RNAs based on scores and return final list (`R`, `crisprVerse`)
5. Generate report with overview figures and statistics (`R markdown`)
6. Return report as HTML and PDF files (`weasyprint`)
7. Export module logs and versions

If you want to contribute, report issues, or suggest features, please get in touch on [github](https://github.com/MPUSP/snakemake-crispr-guides).

----------

# Prerequisites

## Packages

Loaded required R packages:

- `tidyverse`
- `ggrepel`
- `ggpubr`
- `GGally`


```{r, echo = FALSE, warning = FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(ggrepel)
  library(ggpubr)
  library(GGally)
})
```


```{r, echo = FALSE}
# import required parameters
target_type <- snakemake@config$design_guides$target_type
tss_window <- snakemake@config$design_guides$tss_window
tiling_window <- snakemake@config$design_guides$tiling_window
n_examples <- snakemake@config$report$show_examples
show_genomic_range <- snakemake@config$report$show_genomic_range

# custom ggplot2 theme that is reused for all later plots
custom_colors <- c("#E7298A", "#66A61E", "#E6AB02", "#7570B3", "#B3B3B3", "#1B9E77", "#D95F02", "#A6761D")
custom_range <- function(n = 5) {
  colorRampPalette(custom_colors[c(1, 5, 2)])(n)
}

custom_theme <- function(base_size = 12, base_line_size = 1.0, base_rect_size = 1.0, ...) {
  theme_light(base_size = base_size, base_line_size = base_line_size, base_rect_size = base_rect_size) + theme(
    title = element_text(colour = grey(0.4), size = 10),
    plot.margin = unit(c(12, 12, 12, 12), "points"),
    axis.ticks.length = unit(0.2, "cm"),
    axis.ticks = element_line(colour = grey(0.4), linetype = "solid", lineend = "round"),
    axis.text.x = element_text(colour = grey(0.4), size = 10),
    axis.text.y = element_text(colour = grey(0.4), size = 10),
    panel.grid.major = element_line(linewidth = 0.6, linetype = "solid", colour = grey(0.9)),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(linetype = "solid", colour = grey(0.4), fill = NA, linewidth = 1.0),
    panel.background = element_blank(),
    strip.background = element_blank(),
    strip.text = element_text(colour = grey(0.4), size = 10, margin = unit(rep(3, 4), "points")),
    legend.text = element_text(colour = grey(0.4), size = 10),
    legend.title = element_blank(),
    legend.background = element_blank(),
    ...
  )
}

# set graphical parameter for subfigure labels
list_fontpars <- list(face = "plain", size = 14)

# default figure size
figwidth <- 7.0
figheight <- 4.5
figheight2 <- 1.1 + (n_examples * 0.75)

# function to export an image as svg and png
save_plot <- function(pl, path = "../figures/", width = 6, height = 6) {
  pl_name <- deparse(substitute(pl))
  svg(
    filename = paste0(path, pl_name, ".svg"),
    width = width, height = height
  )
  print(pl)
  dev.off()
  png(
    filename = paste0(path, pl_name, ".png"),
    width = width * 125, height = height * 125, res = 120
  )
  print(pl)
  invisible(capture.output(dev.off()))
}
```

## Configuration

Show parameters that were used for guide RNA prediction.
The displayed parameters are taken from the pipelines `config.yml` file.

```{r, echo = FALSE}
# show all parameters for pipeline
list_config <- read_lines("config/config.yml")
list_config <- gsub("\"", "", list_config)
print(list_config)
```

## Result tables {.tabset}

This section shows the imported result tables with predicted guide RNAs.
Only the head of the table is shown (first 10 rows).

*NOTE: tables are only rendered in HTML output but not PDF.*

```{r, echo = FALSE}
input_tx <- snakemake@input$list_tx
list_tx <- read_csv(input_tx, show_col_types = FALSE)

for (tg in target_type) {
  input_file <- paste0("results/filter_guides/guideRNAs_", tg, ".csv")
  assign(paste0("df_", tg), read_csv(input_file, show_col_types = FALSE))
}

if ("df_intergenic" %in% ls()) {
  figheight3 <- 1.1 + (length(unique(df_intergenic$seqnames)) * 0.75)
} else {
  figheight3 <- figheight2
}
```

### Targets (genes) {-}

- table of guide RNAs targeting genes etc.
- targets are defined by a certain window size around the annotated transcritions start site
- defined by variable `target_type` in config file, value `target`

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("df_target" %in% ls()) {
  head(df_target, n = 10)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

- table of guide RNAs targeting intergenic regions
- a tiling approach is used to find suitable guide RNAs in regular intervals
- in the absence of genes, the entire input sequence is tiled
- defined by variable `target_type` in config file, value `intergenic`

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("df_intergenic" %in% ls()) {
  head(df_intergenic, n = 10)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

- table of guide RNAs with no target at all
- random sequences are used to generate these guides
- limited filtering is performed to exclude off-targets and other problems
- defined by variable `target_type` in config file, value `ntc`

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("df_ntc" %in% ls()) {
  head(df_ntc, n = 10)
} else {
  print("No data with guide RNAs for no-target controls available.")
}
```

----------

# Summary statistics

## Number of predicted small guide RNAs {.tabset}

- Overview about number of targeted genes or other elements
- Overview about number of predicted small guide RNAs (sgRNAs)
- Distribution of number of sgRNAs per target (gene)

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  if (tg == "ntc") {
    df_ntc$tx_name <- "no-target"
  }

  df_n_guides <- get(paste0("df_", tg)) %>%
    group_by(tx_name) %>%
    summarize(n_guides = n()) %>%
    count(n_guides)

  plot_n_guides <- df_n_guides %>%
    ggplot(aes(x = n_guides, y = n, label = n)) +
    geom_col(color = NA, fill = custom_colors[1]) +
    geom_text(size = 3.5, nudge_y = max(df_n_guides$n) * 0.03, color = grey(0.5)) +
    labs(
      x = "guides per target", y = "count",
      title = paste0("Number of predicted guide RNAs per target (total N = ", nrow(get(paste0("df_", tg))), ")"),
      subtitle = paste0("Number of targets with predicted guide RNAs: ", length(unique(get(paste0("df_", tg))$tx_name)))
    ) +
    custom_theme()

  assign(paste0("plot_n_guides_", tg), plot_n_guides)
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_n_guides_target" %in% ls()) {
  print(plot_n_guides_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_n_guides_intergenic" %in% ls()) {
  print(plot_n_guides_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_n_guides_ntc" %in% ls()) {
  print(plot_n_guides_ntc)
} else {
  print("No data with guide RNAs for no-target controls available.")
}
```

## Strand, width, position, and PAM site {.tabset}

- strand of gene-targeting guides is one of "+" (template strand) or "-" (non-template strand)
- strand of intergenic regions refers to "+" and "-" strand of genome
- no-target controls do not have meaningful strand information

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  df_guides <- get(paste0("df_", tg))
  if (!"dist_to_tss" %in% colnames(df_guides)) {
    df_guides$dist_to_tss <- 0
  }

  plot_strand <- ggpubr::ggarrange(
    df_guides %>%
      count(strand) %>%
      ggplot(aes(x = factor(strand), y = n, label = n)) +
      geom_col(color = "white", fill = custom_colors[1]) +
      geom_text(size = 3.5, nudge_y = nrow(df_guides) / 30, color = grey(0.5)) +
      labs(x = "strand", y = "count", title = "Guides per strand") +
      custom_theme(),
    df_guides %>%
      count(width) %>%
      ggplot(aes(x = factor(width), y = n, label = n)) +
      geom_col(color = "white", fill = custom_colors[1]) +
      geom_text(size = 3.5, nudge_y = nrow(df_guides) / 30, color = grey(0.5)) +
      labs(x = "width", y = "count", title = "Width distribution of guides") +
      custom_theme(),
    df_guides %>%
      count(pam) %>%
      ggplot(aes(x = factor(pam), y = n, label = n)) +
      geom_col(color = "white", fill = custom_colors[1]) +
      geom_text(size = 3.5, nudge_y = nrow(df_guides) / 30, color = grey(0.5)) +
      labs(x = "pam", y = "count", title = "PAM sequence of guides") +
      custom_theme(),
    df_guides %>%
      ggplot(aes(x = dist_to_tss)) +
      geom_histogram(bins = 30, color = "white", fill = custom_colors[1]) +
      labs(x = "distance [nt]", y = "count", title = "Relative distance to TSS") +
      custom_theme()
  )
  assign(paste0("plot_strand_", tg), plot_strand)
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_strand_target" %in% ls()) {
  print(plot_strand_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_strand_intergenic" %in% ls()) {
  print(plot_strand_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_strand_ntc" %in% ls()) {
  print(plot_strand_ntc)
} else {
  print("No data with guide RNAs for no-target controls available.")
}
```

## On-target scores {.tabset}

- What is the distribution of on-target scores?
- 'Good' distributions should be normally distributed between a range of 0 and 1
- 'Bad' distributions are too narrow, indicating that the score is probably not useful to differentiate between high and low efficacy

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  if (tg != "ntc") {
    df_guides <- get(paste0("df_", tg))
    plot_scores <- df_guides %>%
      select(matches("^score_") & !matches("^score_(cfd|mit)")) %>%
      pivot_longer(
        cols = everything(),
        names_to = "score",
        values_to = "value",
        names_transform = function(x) gsub("score_", "", x)
      ) %>%
      ggplot(aes(x = value)) +
      geom_histogram(bins = 30, color = "white", fill = custom_colors[1]) +
      labs(x = "score", y = "count", title = "Global distribution of on-target scores") +
      facet_wrap(~score) +
      custom_theme()

    assign(paste0("plot_scores_", tg), plot_scores)
  }
}
```


### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_scores_target" %in% ls()) {
  print(plot_scores_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_scores_intergenic" %in% ls()) {
  print(plot_scores_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
print("No data with guide RNAs for no-target controls available.")
```

## Correlation between scores {.tabset}

- What is the correlation between scores?
- Depicted is the scatter plot of each score against each other, as well as the correlation coefficient _R_
- It is not unusual that scores show a low degree of correlation with each other
- However a slight positive correlation should be visible; a negative correlation would mean that scores are contradicting each other

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  if (tg != "ntc") {
    df_guides <- get(paste0("df_", tg))
    plot_scorecorr <- df_guides %>%
      select(matches("^score_") & !matches("^score_(cfd|mit)")) %>%
      rename_with(.fn = function(x) gsub("score_", "", x)) %>%
      ggpairs(
        upper = list(continuous = "cor"),
        lower = list(continuous = wrap("points", alpha = 0.8, size = 0.3, color = custom_colors[1])),
        diag = list(continuous = wrap("densityDiag", color = grey(0.3), fill = custom_colors[1])),
        progress = FALSE
      ) +
      labs(
        x = "", y = "",
        title = "Correlation of scores with each other",
        subtitle = "Lower triangle: scatter plots, upper triangle: correlation coefficent R"
      ) +
      custom_theme()

    assign(paste0("plot_scorecorr_", tg), plot_scorecorr)
  }
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_scorecorr_target" %in% ls()) {
  print(plot_scorecorr_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_scorecorr_intergenic" %in% ls()) {
  print(plot_scorecorr_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
print("No data with guide RNAs for no-target controls available.")
```


## Location of guide RNAs - overview {.tabset}

- The following figures show the position of guide RNAs on a selection of target genes
- The first figure shows the distribution of all relative guide RNA positions

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  df_guides <- get(paste0("df_", tg))
  if (tg == "intergenic") {
    titletext <- "Distribution of all guide RNAs over the tiling windows"
    df_guides <- df_guides %>%
      mutate(dist_to_tss = start - tiling_window * tx_window)
  }
  if (tg == "target") {
    titletext <- "Distribution of all guide RNAs over the TSS window"
  }
  if (tg != "ntc") {
    plot_location <- df_guides %>%
      ggplot(aes(x = dist_to_tss)) +
      geom_histogram(fill = custom_colors[1], bins = 50) +
      geom_smooth(
        data = count(df_guides, dist_to_tss),
        aes(x = dist_to_tss, y = n),
        method = "loess", formula = y ~ x,
        color = grey(0.3), fill = grey(0.6)
      ) +
      labs(
        x = "", y = "",
        title = titletext,
        subtitle = "Bars are sgRNAs counts per position, ribbon is smoothed average"
      ) +
      custom_theme()

    assign(paste0("plot_location_", tg), plot_location)
  }
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_location_target" %in% ls()) {
  print(plot_location_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_location_intergenic" %in% ls()) {
  print(plot_location_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
print("No data with guide RNAs for no-target controls available.")
```

## Location of guide RNAs - examples {.tabset}

- These figures show a selection of the first N genes/regions for the purpose of quality control
- For intergenic regions, the first N genomic windows ("tiles") with width `tiling_window` are shown
- If both genes and intergenic regions are targeted, some intergenic regions might be too narrow to find guides
- The number of example genes/regions is controlled by parameter `show_examples` in the config file

```{r, echo = FALSE}
# function to draw guide RNAs for genes/targets
draw_guide_target <- function(df, n_examples) {
  df %>%
    filter(tx_name %in% unique(df$tx_name)[1:n_examples]) %>%
    mutate(tx_name = paste0(tx_name, " (", tss_strand, ")") %>%
      factor(., unique(.))) %>%
    mutate(
      xstart = case_when(
        tss_strand == "+" & strand == "+" ~ start - tss_pos,
        tss_strand == "+" & strand == "-" ~ end - tss_pos,
        tss_strand == "-" & strand == "-" ~ -1 * (end - tss_pos),
        tss_strand == "-" & strand == "+" ~ -1 * (start - tss_pos),
      ),
      xend = case_when(
        tss_strand == "+" & strand == "+" ~ end - tss_pos,
        tss_strand == "+" & strand == "-" ~ start - tss_pos,
        tss_strand == "-" & strand == "-" ~ -1 * (start - tss_pos),
        tss_strand == "-" & strand == "+" ~ -1 * (end - tss_pos),
      ),
      y = ifelse(strand == tss_strand, 0.25, -0.25)
    ) %>%
    ggplot(aes(x = xstart, y = y, xend = xend, yend = y)) +
    geom_segment(
      x = tss_window[1], y = 0, xend = tss_window[2], yend = 0,
      linewidth = 2, lineend = "square", linejoin = "mitre", color = grey(0.7)
    ) +
    geom_segment(aes(x = 0, y = 0, xend = tx_width, yend = 0),
      linewidth = 1.33, color = grey(0.3), lineend = "square", linejoin = "mitre", arrow = arrow(
        angle = 30, length = unit(0.02, "inches"),
        type = "closed"
      )
    ) +
    geom_segment(aes(color = ifelse(strand == tss_strand, "+", "-")),
      linewidth = 1.5, lineend = "square", linejoin = "mitre", arrow = arrow(
        angle = 30, length = unit(0.01, "inches"),
        type = "closed"
      )
    ) +
    facet_wrap(~tx_name, ncol = 1) +
    coord_cartesian(ylim = c(-0.5, 0.5), xlim = tss_window) +
    labs(
      x = "distance to TSS [nt]", y = "",
      title = paste0("Guide RNA location in TSS window, for the first ", n_examples, " genes"),
      subtitle = paste0(
        "Colored arrows: guide RNAs, gray line: TSS window, black arrow: transcript.\n",
        "Note that genes on the '-' strand are shown as reverse-complement"
      )
    ) +
    custom_theme(legend.position = "none") +
    scale_color_manual(values = custom_colors[1:2]) +
    theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())
}

# for no-gene targeting guides, we plot an overview about the genomic region(s)
draw_guide_intergenic <- function(
    df, n_examples, show_genomic_range = c(0, 50000)) {
  df %>%
    filter(seqnames %in% unique(df$seqnames)[1:n_examples]) %>%
    mutate(
      xstart = ifelse(strand == "+", start, end),
      xend = ifelse(strand == "+", end, start),
      y = ifelse(strand == "+", 0.25, -0.25)
    ) %>%
    ggplot(aes(x = xstart, y = y, xend = xend, yend = y)) +
    geom_segment(
      data = list_tx,
      aes(x = start, y = ifelse(strand == "+", 0.125, -0.125), xend = end, yend = ifelse(strand == "+", 0.125, -0.125)),
      linewidth = 2, lineend = "square", linejoin = "mitre", color = grey(0.7)
    ) +
    geom_segment(aes(color = strand),
      linewidth = 1.5, lineend = "square", linejoin = "mitre", arrow = arrow(
        angle = 30, length = unit(0.01, "inches"),
        type = "closed"
      )
    ) +
    facet_wrap(~seqnames, ncol = 1) +
    coord_cartesian(ylim = c(-0.5, 0.5), xlim = show_genomic_range) +
    labs(
      x = "position [nt]", y = "",
      title = paste0("Guide RNA location per chromosome (max. ", n_examples, "shown)"),
      subtitle = "Colored arrows: guide RNAs, gray lines: annotated genes."
    ) +
    custom_theme(legend.position = "none") +
    scale_color_manual(values = custom_colors[1:2]) +
    theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight2}
if ("target" %in% target_type) {
  draw_guide_target(df_target, n_examples)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight3}
if ("intergenic" %in% target_type) {
  draw_guide_intergenic(df_intergenic, n_examples, show_genomic_range)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
print("No data with guide RNAs for no-target controls available.")
```

## Guides with multiple targets {.tabset}

 - Some guides target several genes/elements within a genome
 - these multi-target guides can be off-targets (not desired)
 - or they can target duplicated genes/element (desired)
 - if multi-target guides are to be included or not can be controlled with the `filter_multi_targets` config parameter (default: `TRUE`)
 - this figure gives an overview about the number of multi-target guides, if they were not filtered

```{r, echo = FALSE, warning = FALSE}
for (tg in target_type) {
  if (tg != "ntc") {
    df_guides <- get(paste0("df_", tg))
    if ("multi_target" %in% colnames(df_guides)) {
      plot_multi_targets <- df_guides %>%
        mutate(multi_target = ifelse(multi_target, "multi-target", "single target")) %>%
        group_by(multi_target) %>%
        count(seqnames) %>%
        ggplot(aes(x = n, y = "count", fill = multi_target, label = n)) +
        geom_col(position = position_dodge2()) +
        geom_text(aes(color = multi_target, x = n + max(n) / 30),
          position = position_dodge(0.9), size = 3.5
        ) +
        labs(
          x = "", y = "",
          title = "Multi-target guides per chromosome",
          subtitle = "Bars are sgRNA counts, color indicates target type"
        ) +
        facet_wrap(~seqnames, ncol = 1) +
        custom_theme(legend.position = "bottom") +
        scale_fill_manual(values = custom_colors[2:1]) +
        scale_color_manual(values = custom_colors[2:1]) +
        theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())

      assign(paste0("plot_multi_targets_", tg), plot_multi_targets)
    }
  }
}
```

### Targets (genes) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_multi_targets_target" %in% ls()) {
  print(plot_multi_targets_target)
} else {
  print("No data with guide RNAs targeting genes etc available.")
}
```

### Intergenic regions {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
if ("plot_multi_targets_intergenic" %in% target_type) {
  print(plot_multi_targets_intergenic)
} else {
  print("No data with guide RNAs targeting intergenic regions available.")
}
```

### No-target controls (NTC) {-}

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
print("No data with guide RNAs for no-target controls available.")
```

## Targets where no guides were designed

- in some cases the design of guides is not possible, for example when:
- the target (gene) is very short, e.g. below 100 bp, and no PAM site can be found
- a gene/CDS has no parent transcript (orphan CDS)
- a CDS is partially or fully overlapping wit another CDS
- a gene is duplicated, leading to the removal of multi-target guides (see previous section)
- the following table and figure shows the targets where no guides were found, if any

```{r, echo = FALSE}
if ("target" %in% target_type) {
  df_fail <- read_csv("results/filter_guides/guideRNAs_target_failed.csv", show_col_types = FALSE)
  if (nrow(df_fail) > 0) {
    df_fail
  } else {
    print("No genes/targets without guide RNAs were found")
  }
} else {
  print("No genes were supplied as targets, therefore no design failures.")
}
```

```{r, echo = FALSE}
if ("target" %in% target_type) {
  if (nrow(df_fail) > 0) {
    df_fail %>%
      ggplot(aes(x = start / 1000, y = 0, color = strand, label = tx_name)) +
      geom_point() +
      geom_text_repel(size = 2.5, max.overlaps = 50) +
      coord_cartesian(ylim = c(-0.5, 0.5)) +
      labs(
        x = "position [kbp]", y = "",
        title = paste0("Location of targets where no guides could be found (", nrow(df_fail), ")"),
        subtitle = paste0("Dots: genes, color: strand")
      ) +
      facet_wrap(~seqnames, ncol = 1) +
      custom_theme(legend.position = "bottom") +
      scale_color_manual(values = custom_colors[1:2]) +
      theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())
  }
} else {
  print("No genes were supplied as targets, therefore no design failures.")
}
```

----------

# About this report

## Pipeline

This report was automatically generated by the [snakemake-crispr-guides](https://github.com/MPUSP/snakemake-crispr-guides) pipeline.

For issues, bugs, and feature requests please use the pipeline's [github page](https://github.com/MPUSP/snakemake-crispr-guides/issues).

For all other feedback, contact the author(s):

- Michael Jahn, PhD (author and maintainer): jahn@mpusp.mpg.de

The pipeline is developed at the [Max Planck Unit for the Science of Pathogens](https://www.mpusp.mpg.de/), Berlin, Germany.

## Packages used in the pipeline

This list shows the fixed dependencies used as `conda` environment variables within the pipeline.

```{r, echo = FALSE, warnings = FALSE, eval = FALSE}
versions <- read_delim("results/versions/log_packages.txt",
  delim = "\n",
  show_col_types = FALSE,
  col_names = "module"
)

print(versions, n = Inf)
```

## Data accessability

The following resources were used to generate this report:

- Guide RNA results table: `results/design_guides/guideRNAs_top.csv`
- config file: `config/config.yml`

## Session Info

Link to the R markdown source that was used to generate this report:

<a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file</a>

Session info (R base and package versions):

```{r, echo = FALSE}
sessionInfo()
```

```{r, echo = FALSE}
write_lines(
  "REPORT: finished writing HTML report successfully",
  file = snakemake@log$path
)
```
